[9/N] [Dispatchable Collectives] Update reduce_scatter with CPU / CUDA implementations

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #87088 Remove ProcessGroupRoundRobin
* #86409 [13/N] Update gather with CPU/CUDA implementations
* #86408 [12/N] Update scatter with CPU/CUDA implementations
* #86407 [11/N] Update all_to_all with CPU/CUDA implementations
* #86368 [10/N] Update barrier with CPU/CUDA implementations
* **#86166 [9/N] [Dispatchable Collectives] Update reduce_scatter with CPU / CUDA implementations**

### Changes
- Updates for the reduce_scatter collective

### Context
https://github.com/pytorch/pytorch/issues/86225

H-Huang committed:
[9/N] [Dispatchable Collectives] Update reduce_scatter with CPU / CUDA implementations

[ghstack-poisoned]

H-Huang committed:
Update on "[9/N] [Dispatchable Collectives] Update reduce_scatter with CPU / CUDA implementations"


### Changes
- Updates for the reduce_scatter collective

### Context
https://github.com/pytorch/pytorch/issues/86225


[ghstack-poisoned]

H-Huang committed:
Update on "[9/N] [Dispatchable Collectives] Update reduce_scatter with CPU / CUDA implementations"


### Changes
- Updates for the reduce_scatter collective

### Context
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D40145699](https://our.internmc.facebook.com/intern/diff/D40145699)

[ghstack-poisoned]

wanchaol commented on test/distributed/test_c10d_common.py:
```diff
@@ -1466,6 +1466,10 @@ def _call_collective_with_varying_tensors(self, backend, collective, *args):
         if collective == dist.all_gather:
             collective([tensor], tensor, *args)
+        elif collective == dist.reduce_scatter:
+            if backend != "gloo":
```
with the dispatchable functionality, what would happen if user call reduce_scatter on CPU?

H-Huang replied:
Sorry, I missed this question. It will call the `reduce_scatter` based on the configured backend for `CPU` (we will set a backend for CPU by default, but users can configure this). This is yet to be implemented in this PR

H-Huang committed:
Update on "[9/N] [Dispatchable Collectives] Update reduce_scatter with CPU / CUDA implementations"


### Changes
- Updates for the reduce_scatter collective

### Context
https://github.com/pytorch/pytorch/issues/86225

Differential Revision: [D40145699](https://our.internmc.facebook.com/intern/diff/D40145699)

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D40145699).

kwen2501 reviewed (approved):
LGTM.

H-Huang committed:
Update on "[9/N] [Dispatchable Collectives] Update reduce_scatter with CPU / CUDA implementations"


### Changes
- Updates for the reduce_scatter collective

### Context
https://github.com/pytorch/pytorch/issues/86225

[ghstack-poisoned]

H-Huang commented:
@pytorchbot merge
